### ISSUE TITLE
Dropout does not give good accuracy

### ISSUE THREAD

**Issue by ndryden** [LLNL/lbann#3]
Training with dropout does not seem to give the good accuracy the dropout paper (and dropout's wide use) would suggest. The dropout paper gives a few experiments on MNIST that I've tried to reproduce:

1. Dropout NN with three hidden layers of 1024 units, logistic activation, 0.8 keep probability on the input layer, 0.5 keep probability on the hidden layers, gives 1.35% error (98.65% accuracy)
2. Ditto, with ReLU activation gives 1.25% error (98.75% accuracy)
3.  They also do an evaluation (see figure 4, their JMLR paper) in which they train a NN (e.g. three hidden layers, 1024 units in each) without dropout, then use the exact same hyperparameters but with dropout on the hidden layers. Without dropout, error is ~1.7-1.8%; with dropout error is ~1.0-1.3%.

Trying these same experiments in LBANN, our ReLU accuracy without dropout is comparable (~98.3%). Our sigmoid accuracy without dropout is ~96.8% (this is low due to using the same learning rate as with ReLU-- we can do better). But with dropout, our accuracy is actually worse:
* ReLU with 0.5 dropout on the hidden layers gets ~97.0% accuracy.
* Sigmoid with 0.5 dropout on the hidden layers gets ~94.0% accuracy.
* Adding dropout to the input layer just makes things worse.
(These all used a learning rate 0.005.)

Based on this, I suspect there's a bug somewhere in dropout/

**Comment by timmoon10** [comment]
Dropout now checks whether it is training and disables itself at test time. Does this fix the issue?

I warn that this commit includes a bug. When I apply dropout with p=0.9999 to the hidden layer between a pooling layer and a fully-connected layer, we take a big hit in accuracy. This issue disappears when I don't scale the dropout mask by 1/p.

**Comment by ndryden** [comment]
I ran some models in Keras, including their default MNIST example and the model I ran above. We now get comparable accuracy. It's still a bit less than what the dropout paper reports, but this could be attributed to them selecting better hyperparameters and training for longer (they used 1 million weight updates to select hyperparameters, then trained the best one for 1 million weight updates).

For reference, we still get 98.3% accuracy without dropout. A 784-1024-1024-1024-10 network with ReLU activations and 0.8 dropout on the input and fully-connected layers seems to converge to 98.35%.
